api/circulation.py

```diff
@@ -601,76 +599,81 @@ def borrow(self, patron, pin, licensepool, delivery_mechanism,
 
         new_loan = False
 
-        loan_limit = patron.library.setting(Configuration.LOAN_LIMIT).int_value
-        non_open_access_loans_with_end_date = [loan for loan in patron.loans if loan.license_pool.open_access == False and loan.end]
-        at_loan_limit = (loan_limit and len(non_open_access_loans_with_end_date) >= loan_limit)
-
-        # If we're at the loan limit, skip trying to check out the book and just try
-        # to place a hold. Otherwise, try to check out the book even if we think it's
-        # not available.
-        if not at_loan_limit:
-            try:
-                loan_info = api.checkout(
-                    patron, pin, licensepool, internal_format
-                )
+        # Enforce any library-specific limits on loans or holds.
+        self.enforce_limits(patron, licensepool)
+
+        # Since that didn't raise an exception, we know that the
+        # patron is able to get a loan or a hold. There are race
+        # conditions that will allow someone to get a hold in excess
+        # of their hold limit (because we thought they were getting a
+        # loan but someone else checked out the book right before we
+        # got to it) but they're rare and not serious.
+
+        # We try to check out the book even if we believe it's not
+        # available -- someone else may have checked it in since we
+        # last looked.
+        try:
+            loan_info = api.checkout(
+                patron, pin, licensepool, internal_format
+            )
 
-                if isinstance(loan_info, HoldInfo):
```

Comment:
For the most part, this code wasn't removed; it was dedented because the `if not at_loan_limit:` if statement was removed.